coding/javascript.md

```diff
@@ -8,25 +8,160 @@ redirect_from:
 
 This document outlines the rules for writing JavaScript across our codebase.
 
-## Progressive Enhancement
+## General Principles
+
+### Progressive Enhancement
 
 JavaScript should progressively enhance a page.
 
 On all publicly accessible websites, core content and functionality should still be available without JavaScript. JavaScript applications written for a more specific audience which won't be indexed by search engines may be more lax, but should still treat progressive enhancement as a guiding principle.
 
-## Performance
+### Performance
 
 - Cache DOM queries — only select an element once.
 - Use event delegation as much as possible to reduce the number of events bound to the page
 - Understand when you cause an element repaint and reduce inline style manipulations accordingly.
 - If possible, always use CSS for element transforms and transitions.
 
-## Element hooks
+### Element hooks
 
 We should target page element in JavaScript using the `data-js` element attribute instead of using the `class` attribute. Classnames should be used for styling only.
 
 e.g. `<a data-js="index-link" href="/index">Index</a>`.
 
+## TypeScript
+
+### TypeScript vs JavaScript
+
+For any medium-large JavaScript applications, consider using TypeScript in preference to JavaScript, as it offers a better developer experience, and increased safety. See [Transitioning to TypeScript](https://docs.google.com/document/d/1yXNI_biuKwoHBk2goLEksZayV_MGAkQxukvSNLCDtnE/edit#heading=h.yij8xaij9lcy) for more details.
+
+Remember that TypeScript gives you the ability to _progressively_ type your project; it isn't an all or nothing commitment. If you like, you can convert JavaScript files to TypeScript one at a time, if migrating an existing project.
+
+TypeScript is currently used in the following projects:
+
+- [maas-ui](https://github.com/canonical-web-and-design/react-components/)
+- [react-components](https://github.com/canonical-web-and-design/react-components/)
+
+### Patterns
```

Comment:
These are really helpful, thanks!

Maybe it would be also worth adding some links to real code examples?
Ideally for each pattern we could have "here is how we use it in this component" link. Or maybe just "Some real TS component examples" section with a couple examples?

Reply:
Good suggestion! I can add a link to some real-world examples in `react-components`.

Reply:
Done!